Add support for requestBody to webRequest.onBeforeRequest.

xeenon opened the pull request:
#### e06024877fb21a6b7c7604fedd662d5fe9a249ed
<pre>
Add support for requestBody to webRequest.onBeforeRequest.
<a href="https://webkit.org/b/285796">https://webkit.org/b/285796</a>
<a href="https://rdar.apple.com/140338580">rdar://140338580</a>

Reviewed by Brian Weinstein.

The `requestBody` property has never been supported correctly (it was missing the `formData` object).
This stopped working altogether when moved to `WebCore::ResourceRequest`, because the HTTP body
is not serialized with the request.

Fix this by sending `IPC::FormDataReference` as well, and parse the data with `URLParser` to build
the `formData` object key-value pairs. Also add support for the `raw` array with `TypedArray` values.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::resourceLoadDidSendRequest):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::toWebAPI): Added.
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest):
(WebKit::convertRequestBodyToWebExtensionFormat): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::enumerateListeners):
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebRequestEvent.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndFormData)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndBlob)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndJSON)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndForm)): Added.

Canonical link: <a href="https://commits.webkit.org/288813@main">https://commits.webkit.org/288813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f06d592f2a5300bee4137124cd4719dd410a9a8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84346 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3971 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38651 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89425 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35355 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86431 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4056 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65597 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23436 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87392 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76636 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45891 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2997 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30866 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34404 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31635 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90806 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8440 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74047 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73249 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17578 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16022 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2987 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13084 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11565 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17041 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11414 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14890 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13187 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
```diff
@@ -233,21 +279,29 @@
+
+            listener.call(details.get());
+
+            [details removeObjectForKey:@"requestBody"];
```
Nit: Should `requestBody` be in a constant somewhere?

xeenon replied:
We didn't use constants for some reaosn in this file, so I kept with that. We should consider converting all the strings in here to constants. I have another PR coming that touches webRequest, so I will do that in a follow up to avoid merge issues.

xeenon committed:
Add support for requestBody to webRequest.onBeforeRequest.
https://webkit.org/b/285796
rdar://140338580

Reviewed by Brian Weinstein.

The `requestBody` property has never been supported correctly (it was missing the `formData` object).
This stopped working altogether when moved to `WebCore::ResourceRequest`, because the HTTP body
is not serialized with the request.

Fix this by sending `IPC::FormDataReference` as well, and parse the data with `URLParser` to build
the `formData` object key-value pairs. Also add support for the `raw` array with `TypedArray` values.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::resourceLoadDidSendRequest):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::toWebAPI): Added.
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest):
(WebKit::convertRequestBodyToWebExtensionFormat): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::enumerateListeners):
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebRequestEvent.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndFormData)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndBlob)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndJSON)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, BeforeRequestEventWithRequestBodyAndForm)): Added.

Canonical link: https://commits.webkit.org/288813@main
